From 719479c31a7921c422fc980c9c91585f5e01b5ec Mon Sep 17 00:00:00 2001
From: Mohan Kumar G <quic_mkumarg@quicinc.com>
Date: Fri, 29 Nov 2024 09:53:32 +0530
Subject: [PATCH] UPSTREAM:PROTOCOL:wifi: ath12k: Send MCS15 support to FW
 during peer assoc

Send mcs15 support to FW through WMI command during peer
assoc.

Patch-depdendency: None

Patch-work: None

Feature: MCS15

Signed-off-by: Mohan Kumar G <quic_mkumarg@quicinc.com>
Signed-off-by: Dhanavandhana Kannan <quic_dhanavan1@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 7 +++++++
 drivers/net/wireless/ath/ath12k/wmi.c | 3 +++
 drivers/net/wireless/ath/ath12k/wmi.h | 1 +
 3 files changed, 11 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -3758,6 +3758,7 @@ static void ath12k_peer_assoc_h_eht(stru
 	const struct ieee80211_sta_eht_cap *eht_cap;
 	const struct ieee80211_eht_mcs_nss_supp_20mhz_only *bw_20;
 	const struct ieee80211_eht_mcs_nss_supp_bw *bw;
+	struct ieee80211_bss_conf *link_conf;
 	u32 *rx_mcs, *tx_mcs;
 
 	enum nl80211_band band;
@@ -3789,6 +3790,11 @@ static void ath12k_peer_assoc_h_eht(stru
 		return;
 	}
 
+	link_conf = ath12k_get_link_bss_conf(arvif);
+	if (!link_conf) {
+		ath12k_warn(ar->ab, "unable to access link_conf in peer assoc eht set\n");
+		return;
+	}
 	sta = container_of((void *)ahsta, struct ieee80211_sta, drv_priv);
 
 	eht_cap = &link_sta->eht_cap;
@@ -3938,6 +3944,7 @@ static void ath12k_peer_assoc_h_eht(stru
 
 	max_nss = min(max_nss, (uint8_t)eht_nss);
 
+	arg->enable_mcs15 = link_conf->enable_mcs15;
 	arg->peer_nss = min(link_sta->rx_nss, max_nss);
 	arg->punct_bitmap = ~def.ru_punct_bitmap;
 	if (ieee80211_vif_is_mesh(vif) && link_sta->ru_punct_bitmap)
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -3470,6 +3470,9 @@ int ath12k_wmi_send_peer_assoc_cmd(struc
 				 arg->ml.num_partner_links * sizeof(*partner_info));
 	ptr += TLV_HDR_SIZE;
 
+	if (!arg->enable_mcs15)
+		cmd->peer_eht_ops |= IEEE80211_EHT_OPER_MCS15_DISABLE;
+
 	for (i = 0; i < arg->ml.num_partner_links; i++) {
 		partner_info = ptr;
 		partner_info->tlv_header = FIELD_PREP(WMI_TLV_TAG,
--- a/drivers/net/wireless/ath/ath12k/wmi.h
+++ b/drivers/net/wireless/ath/ath12k/wmi.h
@@ -4672,6 +4672,7 @@ struct ath12k_wmi_peer_assoc_arg {
 	u32 punct_bitmap;
 	bool is_assoc;
 	struct peer_assoc_mlo_params ml;
+	bool enable_mcs15;
 };
 
 struct wmi_peer_assoc_mlo_partner_info {
